File: sdks/typescript/tests/security-profile-compat.test.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import { JsonMap } from "../src/jsonSupport.js";
import { defaultAgentOptions, optionsFromConfigMap, optionsToConfigMap } from "../src/options.js";

const VECTORS_DIR = join(process.cwd(), "..", "tests", "vectors", "security");

function loadFixture(name: string): JsonMap {
  return JSON.parse(readFileSync(join(VECTORS_DIR, name), "utf-8")) as JsonMap;
}

describe("Security profile compatibility", () => {
  it("reads shared security https fixture with expected schema", () => {
    const fixture = loadFixture("security_profile_https.json");
    const options = optionsFromConfigMap(fixture);
    expect(options.key_provider).toBe("vault");
    expect(options.vault_url).toBe("https://vault.company.net");
    expect(options.vault_path).toBe("secret/data/acp/identities");
    expect(options.vault_token_env).toBe("VAULT_TOKEN");
    expect(options.allow_insecure_http).toBe(false);
    expect(options.allow_insecure_tls).toBe(false);
    expect(options.mtls_enabled).toBe(false);
  });

  it("reads shared security vault + mtls fixture", () => {
    const fixture = loadFixture("security_profile_vault_mtls.json");
    const options = optionsFromConfigMap(fixture);
    expect(options.key_provider).toBe("vault");
    expect(options.mtls_enabled).toBe(true);
    expect(options.ca_file).toBe("/etc/acp/ca/security-profile-ca.pem");
    expect(options.cert_file).toBeUndefined();
    expect(options.key_file).toBeUndefined();
  });

  it("exports aligned security profile fields", () => {
    const options = {
      ...defaultAgentOptions(),
      key_provider: "vault" as const,
      vault_url: "https://vault.company.net",
      vault_path: "secret/data/acp/identities",
      vault_token_env: "VAULT_TOKEN",
      allow_insecure_http: false,
      allow_insecure_tls: false,
      mtls_enabled: true,
      ca_file: "/etc/acp/ca/security-profile-ca.pem"
    };
    const exported = optionsToConfigMap(options);
    expect(exported.key_provider).toBe("vault");
    expect(exported.vault_url).toBe("https://vault.company.net");
    expect(exported.vault_path).toBe("secret/data/acp/identities");
    expect(exported.vault_token_env).toBe("VAULT_TOKEN");
    expect(exported.allow_insecure_http).toBe(false);
    expect(exported.allow_insecure_tls).toBe(false);
    expect(exported.mtls_enabled).toBe(true);
    expect(exported.ca_file).toBe("/etc/acp/ca/security-profile-ca.pem");
  });
});

File: sdks/go/security_profile_compat_test.go
package acp

import (
	"path/filepath"
	"testing"
)

func securityVectorsDir(t *testing.T) string {
	t.Helper()
	return filepath.Join("..", "tests", "vectors", "security")
}

func TestReadsSharedSecurityHTTPSFixtureWithExpectedSchema(t *testing.T) {
	fixture := loadJSONFixtureMap(t, filepath.Join(securityVectorsDir(t), "security_profile_https.json"))
	options := OptionsFromConfigMap(fixture)
	if options.KeyProvider != "vault" {
		t.Fatalf("key_provider mismatch: %s", options.KeyProvider)
	}
	if options.VaultURL != "https://vault.company.net" {
		t.Fatalf("vault_url mismatch: %s", options.VaultURL)
	}
	if options.VaultPath != "secret/data/acp/identities" {
		t.Fatalf("vault_path mismatch: %s", options.VaultPath)
	}
	if options.VaultTokenEnv != "VAULT_TOKEN" {
		t.Fatalf("vault_token_env mismatch: %s", options.VaultTokenEnv)
	}
	if options.AllowInsecureHTTP {
		t.Fatalf("allow_insecure_http must be false")
	}
	if options.AllowInsecureTLS {
		t.Fatalf("allow_insecure_tls must be false")
	}
	if options.MTLSEnabled {
		t.Fatalf("mtls_enabled must be false")
	}
}

func TestReadsSharedSecurityVaultMTLSFixtureWithProviderBackedMaterial(t *testing.T) {
	fixture := loadJSONFixtureMap(t, filepath.Join(securityVectorsDir(t), "security_profile_vault_mtls.json"))
	options := OptionsFromConfigMap(fixture)
	if options.KeyProvider != "vault" {
		t.Fatalf("key_provider mismatch: %s", options.KeyProvider)
	}
	if !options.MTLSEnabled {
		t.Fatalf("mtls_enabled must be true")
	}
	if options.CAFile != "/etc/acp/ca/security-profile-ca.pem" {
		t.Fatalf("ca_file mismatch: %s", options.CAFile)
	}
	if options.CertFile != "" {
		t.Fatalf("cert_file should remain empty for provider-backed cert material")
	}
	if options.KeyFile != "" {
		t.Fatalf("key_file should remain empty for provider-backed key material")
	}
}

func TestToConfigMapExportsAlignedSecurityFields(t *testing.T) {
	options := DefaultAgentOptions()
	options.KeyProvider = "vault"
	options.VaultURL = "https://vault.company.net"
	options.VaultPath = "secret/data/acp/identities"
	options.VaultTokenEnv = "VAULT_TOKEN"
	options.AllowInsecureHTTP = false
	options.AllowInsecureTLS = false
	options.MTLSEnabled = true
	options.CAFile = "/etc/acp/ca/security-profile-ca.pem"
	exported := options.ToConfigMap()

	if exported["key_provider"] != "vault" {
		t.Fatalf("key_provider export mismatch")
	}
	if exported["vault_url"] != "https://vault.company.net" {
		t.Fatalf("vault_url export mismatch")
	}
	if exported["vault_path"] != "secret/data/acp/identities" {
		t.Fatalf("vault_path export mismatch")
	}
	if exported["vault_token_env"] != "VAULT_TOKEN" {
		t.Fatalf("vault_token_env export mismatch")
	}
	if exported["allow_insecure_http"] != false {
		t.Fatalf("allow_insecure_http export mismatch")
	}
	if exported["allow_insecure_tls"] != false {
		t.Fatalf("allow_insecure_tls export mismatch")
	}
	if exported["mtls_enabled"] != true {
		t.Fatalf("mtls_enabled export mismatch")
	}
	if exported["ca_file"] != "/etc/acp/ca/security-profile-ca.pem" {
		t.Fatalf("ca_file export mismatch")
	}
}

func TestFromConfigMapPreservesDefaultProviderValuesWhenUnset(t *testing.T) {
	options := OptionsFromConfigMap(map[string]any{})
	if options.KeyProvider != "local" {
		t.Fatalf("default key_provider mismatch")
	}
	if options.VaultTokenEnv != "VAULT_TOKEN" {
		t.Fatalf("default vault_token_env mismatch")
	}
}
